include/exadg/incompressible_navier_stokes/spatial_discretization/spatial_operator_base.cpp

```diff
@@ -249,25 +249,101 @@ template<int dim, typename Number>
-    dealii::Utilities::pow(param.degree_u + 1, dim) * dim;
+  if(param.spatial_discretization == SpatialDiscretization::HDIV)
+  {
+    // We need to make sure the normal dofs are shared between cells on the periodic boundaries,
+    // since these are continuous for HDIV.
+    dealii::IndexSet relevant_dofs;
+    dealii::DoFTools::extract_locally_relevant_dofs(dof_handler_u, relevant_dofs);
+    constraint_u.reinit(relevant_dofs);
+
+    auto const & periodic_face_pair = grid->periodic_faces;
+    for(auto const & face : periodic_face_pair)
+      dealii::DoFTools::make_periodicity_constraints(
+        dof_handler_u,
+        face.cell[0]->face(face.face_idx[0])->boundary_id(),
+        face.cell[1]->face(face.face_idx[1])->boundary_id(),
+        face.face_idx[0] / 2,
+        constraint_u);
+
+
+    // Set boundaries constraints
+    if(boundary_descriptor->velocity->symmetry_bc.empty() == false)
+    {
+      for(auto bc : boundary_descriptor->velocity->symmetry_bc)
+        dealii::VectorTools::project_boundary_values_div_conforming(
+          dof_handler_u, 0, *(bc.second), bc.first, constraint_u, *get_mapping());
+    }
+    if(boundary_descriptor->velocity->dirichlet_bc.empty() == false)
+    {
+      AssertThrow(
+        false,
+        dealii::ExcMessage(
+          "Dirichlet BC are not properly implemented for HDIV. The BC needs to be strongly applied in order to obtain an exactly divergence-free solution."));
+      // We would like to do something similar to above. Probably with
+      // dealii::VectorTools::interpolate_boundary_values
+      // dealii::VectorTools::project_boundary_values
+      // dealii::DoFTools::make_zero_boundary_constraints
+    }
+  }
+
+  unsigned int ndofs_per_cell_velocity;
+  if(get_spatial_discretization() == SpatialDiscretization::L2)
+  {
+    ndofs_per_cell_velocity = dealii::Utilities::pow(param.degree_u + 1, dim) * dim;
+  }
+  else if(get_spatial_discretization() == SpatialDiscretization::HDIV)
+  {
+    ndofs_per_cell_velocity =
+      dealii::Utilities::pow(param.degree_u, dim - 1) * (param.degree_u + 1) * dim;
+  }
+  else
+  {
+    AssertThrow(false, dealii::ExcMessage("FE not implemented."));
+  }
+
   unsigned int const ndofs_per_cell_pressure =
     dealii::Utilities::pow(param.get_degree_p(param.degree_u) + 1, dim);
 
-  pcout << std::endl
-        << "Discontinuous Galerkin finite element discretization:" << std::endl
-        << std::endl
-        << std::flush;
 
   pcout << "Velocity:" << std::endl;
-  print_parameter(pcout, "degree of 1D polynomials", param.degree_u);
+  if(get_spatial_discretization() == SpatialDiscretization::L2)
+  {
+    print_parameter(pcout, "degree of 1D polynomials", param.degree_u);
+  }
+  else if(get_spatial_discretization() == SpatialDiscretization::HDIV)
+  {
+    print_parameter(pcout, "degree of 1D polynomials (normal direction)", param.degree_u);
+    print_parameter(pcout, "degree of 1D polynomials (tangential direction)", (param.degree_u - 1));
```

Comment:
Looks like this needs more characters than the maximum number of characters "reserved" here:
https://github.com/exadg/exadg/blob/8382b8bacb87b907111ffe60e9cc48d7e86c2a4c/include/exadg/utilities/print_functions.h#L70
As a result, the alignment of parameters and their values when printing them to screen will probably no longer work. In my comment here (https://github.com/exadg/exadg/pull/233#discussion_r905978234) you can find another suggestion, or just remove "direction" above.